channel/erc20_depositor.go

```diff
@@ -40,45 +45,134 @@ const ERC20DepositorTXGasLimit = 100000
 func (d *ERC20Depositor) Deposit(ctx context.Context, req DepositReq) (types.Transactions, error) {
-	// Bind a `AssetHolderERC20` instance.
-	assetholder, err := assetholdererc20.NewAssetholdererc20(req.Asset.EthAddress(), req.CB)
-	if err != nil {
-		return nil, errors.Wrapf(err, "binding AssetHolderERC20 contract at: %x", req.Asset)
-	}
+	lockKey := lockKey(req.Account.Address, req.Asset.EthAddress())
+	lock := handleLock(lockKey)
+
 	// Bind an `ERC20` instance.
 	token, err := peruntoken.NewPeruntoken(d.Token, req.CB)
 	if err != nil {
 		return nil, errors.Wrapf(err, "binding ERC20 contract at: %x", d.Token)
 	}
-	// Increase the allowance.
-	opts, err := req.CB.NewTransactor(ctx, ERC20DepositorTXGasLimit, req.Account)
-	if err != nil {
-		return nil, errors.WithMessagef(err, "creating transactor for asset: %x", req.Asset)
+	callOpts := bind.CallOpts{
+		Pending: false,
+		Context: ctx,
+	}
+	// variables for the return value.
+	var depResult DepositResult
+	txApproval, approvalReceived, errApproval := Approve(ctx, lock, req, token, callOpts)
+	if errApproval != nil {
+		depResult.Error = errApproval
+	}
+	if approvalReceived {
+		txDeposit, err := d.DepositOnly(ctx, req)
+		depResult.Transactions = []*types.Transaction{txApproval, txDeposit}
+		depResult.Error = errors.WithMessage(err, "AssetHolderERC20 depositing")
+	} else {
+		depResult.Error = errors.WithMessage(errApproval, "PerunToken was not approved")
 	}
-	tx1, err := token.IncreaseAllowance(opts, req.Asset.EthAddress(), req.Balance)
+	return depResult.Transactions, depResult.Error
+}
+
+// DepositOnly deposits ERC20 tokens into the ERC20 AssetHolder specified at the
+// requests asset address.
+func (d *ERC20Depositor) DepositOnly(ctx context.Context, req DepositReq) (*types.Transaction, error) {
+	// Bind a `AssetHolderERC20` instance.
+	assetholder, err := assetholdererc20.NewAssetholdererc20(req.Asset.EthAddress(), req.CB)
 	if err != nil {
-		err = cherrors.CheckIsChainNotReachableError(err)
-		return nil, errors.WithMessagef(err, "increasing allowance for asset: %x", req.Asset)
+		return nil, errors.Wrapf(err, "binding AssetHolderERC20 contract at: %x", req.Asset)
 	}
 	// Deposit.
-	opts, err = req.CB.NewTransactor(ctx, ERC20DepositorTXGasLimit, req.Account)
+	opts, err := req.CB.NewTransactor(ctx, ERC20DepositorTXGasLimit, req.Account)
 	if err != nil {
 		return nil, errors.WithMessagef(err, "creating transactor for asset: %x", req.Asset)
 	}
+
 	tx2, err := assetholder.Deposit(opts, req.FundingID, req.Balance)
-	err = cherrors.CheckIsChainNotReachableError(err)
-	return []*types.Transaction{tx1, tx2}, errors.WithMessage(err, "AssetHolderERC20 depositing")
+	return tx2, err
 }
 
 // NumTX returns 2 since it does IncreaseAllowance and Deposit.
 func (*ERC20Depositor) NumTX() uint32 {
 	return erc20DepositorNumTx
 }
+
+// Create key from account address and asset to only lock the process when hub deposits the same asset at the same time.
```

Comment:
```suggestion
// Create a lock key from the account address and the asset to ensure only one deposit for an asset is performed at the same time.
```

Note: The Perun Ethereum Backend doesn't know the "hub". And please also correct the formulation for the other docs.